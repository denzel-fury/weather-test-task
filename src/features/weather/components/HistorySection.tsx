import React from 'react'
import { useTranslation } from "react-i18next";

interface HistorySectionProps {
  history: string[];
  deletedItem: string | null;
  onCitySelect: (city: string) => void;
  onDelete: (city: string) => void;
  onUndo: () => void;
}

const HistorySection: React.FC<HistorySectionProps> = ({ history, deletedItem, onCitySelect, onDelete, onUndo }) => {
  const { t } = useTranslation();
  return (
    <div className="mb-4" data-testid="history-section">
      <h3 className="font-bold mb-2">{t("searchHistory")}</h3>
      <ul>
        {history.map((item) => (
          <li key={item} className="flex justify-between items-center mb-1">
            <button
              className="text-blue-600 hover:underline"
              onClick={() => onCitySelect(item)}
              data-testid={`history-${item}`}
            >
              {item}
            </button>
            <button
              onClick={() => onDelete(item)}
              className="text-red-500"
              data-testid={`delete-${item}`}
            >
              X
            </button>
          </li>
        ))}
      </ul>
      {deletedItem && (
        <button
          onClick={onUndo}
          className="text-sm text-green-600 mt-2"
          data-testid="undo-button"
        >
          {t("undoDelete")}
        </button>
      )}
    </div>
  );
};

const MemoizedHistorySection = React.memo(HistorySection);

export default MemoizedHistorySection