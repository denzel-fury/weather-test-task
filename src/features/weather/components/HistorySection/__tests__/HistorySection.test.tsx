import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import HistorySection from '../index';

describe('HistorySection', () => {
  const mockHistory = ['Paris', 'Berlin', 'London'];
  const mockOnCitySelect = jest.fn();
  const mockOnDelete = jest.fn();
  const mockOnUndo = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders history items and calls onCitySelect', () => {
    render(
      <HistorySection
        history={mockHistory}
        deletedItem={null}
        onCitySelect={mockOnCitySelect}
        onDelete={mockOnDelete}
        onUndo={mockOnUndo}
      />
    );

    expect(screen.getByTestId('history-section')).toBeInTheDocument();

    mockHistory.forEach((city) => {
      const cityBtn = screen.getByTestId(`history-${city}`);
      expect(cityBtn).toBeInTheDocument();

      fireEvent.click(cityBtn);
      expect(mockOnCitySelect).toHaveBeenCalledWith(city);
    });
  });

  test('calls onDelete when delete button is clicked', () => {
    render(
      <HistorySection
        history={mockHistory}
        deletedItem={null}
        onCitySelect={mockOnCitySelect}
        onDelete={mockOnDelete}
        onUndo={mockOnUndo}
      />
    );

    const deleteBtn = screen.getByTestId(`delete-Paris`);
    fireEvent.click(deleteBtn);

    expect(mockOnDelete).toHaveBeenCalledWith('Paris');
  });

  test('renders undo button when deletedItem exists and triggers onUndo', () => {
    render(
      <HistorySection
        history={mockHistory}
        deletedItem="Berlin"
        onCitySelect={mockOnCitySelect}
        onDelete={mockOnDelete}
        onUndo={mockOnUndo}
      />
    );

    const undoBtn = screen.getByTestId('undo-button');
    expect(undoBtn).toBeInTheDocument();

    fireEvent.click(undoBtn);
    expect(mockOnUndo).toHaveBeenCalled();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <HistorySection
        history={mockHistory}
        deletedItem="Berlin"
        onCitySelect={mockOnCitySelect}
        onDelete={mockOnDelete}
        onUndo={mockOnUndo}
      />
    );;
    expect(asFragment()).toMatchSnapshot();
  });
});
