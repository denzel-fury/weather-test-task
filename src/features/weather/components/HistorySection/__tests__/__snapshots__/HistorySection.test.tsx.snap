// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`HistorySection matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mb-4"
    data-testid="history-section"
  >
    <h3
      class="font-bold mb-2"
    >
      searchHistory
    </h3>
    <ul>
      <li
        class="flex justify-between items-center mb-1"
      >
        <button
          class="text-blue-600 hover:underline"
          data-testid="history-Paris"
        >
          Paris
        </button>
        <button
          class="text-red-500"
          data-testid="delete-Paris"
        >
          X
        </button>
      </li>
      <li
        class="flex justify-between items-center mb-1"
      >
        <button
          class="text-blue-600 hover:underline"
          data-testid="history-Berlin"
        >
          Berlin
        </button>
        <button
          class="text-red-500"
          data-testid="delete-Berlin"
        >
          X
        </button>
      </li>
      <li
        class="flex justify-between items-center mb-1"
      >
        <button
          class="text-blue-600 hover:underline"
          data-testid="history-London"
        >
          London
        </button>
        <button
          class="text-red-500"
          data-testid="delete-London"
        >
          X
        </button>
      </li>
    </ul>
    <button
      class="text-sm text-green-600 mt-2"
      data-testid="undo-button"
    >
      undoDelete
    </button>
  </div>
</DocumentFragment>
`;
